### PR TITLE
add druid dialect with new pagination query structure

### PR DESCRIPTION
**Pull request by Chandrasekar-Rajasekar**
## Description
Create Druid dialect in aggregation datastore. 

## How Has This Been Tested?
unit test

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Review by aklish — APPROVED** [review]

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java [hunk 1]
```diff
@@ -398,15 +398,20 @@ private SQLQuery toPageTotalSQL(Query query, SQLQuery sql, SQLDialect sqlDialect
...
-        String projectionClause = sqlDialect.generateCountDistinctClause(groupByDimensions);
-
-        return SQLQuery.builder()
-                .projectionClause(projectionClause)
+        SQLQuery innerQuery =  SQLQuery.builder()
+                .projectionClause(groupByDimensions)
                 .fromClause(sql.getFromClause())
                 .joinClause(sql.getJoinClause())
                 .whereClause(sql.getWhereClause())
+                .groupByClause(String.format("GROUP BY %s", groupByDimensions))
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java:406]: Is there any reason we can't simply do:
.groupByClause(sql.getGroupByClause)
and avoid this extra bit of code here?
What if there are no group by dimensions?

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java [hunk 2]
```diff
@@ -389,21 +387,12 @@ private void supplyFilterQueryParameters(Query query, NamedParamPreparedStatemen
...
-        String groupByDimensions =
-                query.getAllDimensionProjections()
-                        .stream()
-                        .map(SQLColumnProjection.class::cast)
-                        .map((column) -> column.toSQL(queryReferenceTable))
-                        .collect(Collectors.joining(", "));
-
         SQLQuery innerQuery =  SQLQuery.builder()
-                .projectionClause(groupByDimensions)
+                .projectionClause(sql.getProjectionClause())
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java:391]: Thinking about this, we ought to remove the metrics if possible or replace them with count or something simple.  Imagine if the metric was a data sketch aggregation.  We don't want to do some expensive operation.

**Comment by Chandrasekar-Rajasekar** [comment]
Couldn't use `groupByClause(sql.getGroupByClause)` for pagination because the main SQL query will not contain group by when a user queries the only dimension and will be in the form 
```
select distinct dims from table 
```
We are generating group by queries for the above cases in pagination.

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java [hunk 3]
```diff
@@ -297,7 +303,8 @@ public String getTableVersion(Table table, Transaction transaction) {
 
         Pagination pagination = query.getPagination();
         if (returnPageTotals(pagination)) {
-            queries.add(toPageTotalSQL(expandedQuery, sql, dialect).toString());
+            SQLQuery paginationSql = toPageTotalSQL(expandedQuery, sql, dialect);
+            queries.add(paginationSql == null ? "" : paginationSql.toString());
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryEngine.java:307]: We don't want to add a query here if we are not running one.  If we return "1" as the rowcount and are not running a query, we'll want to skip adding to the list of queries.

**Review by aklish — CHANGES_REQUESTED** [review]

**Review by aklish — APPROVED** [review]

**Pull request merged**